src/coprocessor/endpoint.rs

```diff
@@ -250,149 +382,247 @@ impl ReqContext {
+    handle_time: Option<f64>,
+
+    region_id: u64,
+    txn_start_ts: u64,
+    ranges_len: usize,
+    first_range: Option<KeyRange>,
+    scan_tag: &'static str,
+    pri_str: &'static str,
+}
+
+impl RequestTracker {
+    fn attach_task_count(&mut self, running_task_count: Arc<AtomicUsize>) {
+        running_task_count.fetch_add(1, Ordering::Release);
+        self.running_task_count = Some(running_task_count);
+    }
+
+    fn attach_ctx_pool(&mut self, ctx_pool: CopContextPool) {
+        self.ctx_pool = Some(ctx_pool);
+    }
+
+    // This function will be only called in thread pool.
+    fn get_basic_metrics(&self) -> RefMut<BasicLocalMetrics> {
+        let ctx_pool = self.ctx_pool.as_ref().unwrap();
+        let ctx = ctx_pool.get_context().0.borrow_mut();
+        RefMut::map(ctx, |c| &mut c.basic_local_metrics)
+    }
+
+    // This function will be only called in thread pool.
+    fn record_wait(&mut self) {
+        let stop_first_wait = self.wait_time.is_none();
+        let wait_start = self.wait_start.take().unwrap();
+        let now = Instant::now_coarse();
+        self.wait_time = Some(duration_to_sec(now - wait_start));
+        self.handle_start = Some(now);
+
+        if stop_first_wait {
+            let ctx_pool = self.ctx_pool.as_ref().unwrap();
+            let mut cop_ctx = ctx_pool.get_context().0.borrow_mut();
+            cop_ctx
+                .basic_local_metrics
+                .wait_time
+                .with_label_values(&[self.scan_tag])
+                .observe(self.wait_time.unwrap());
+        }
+    }
+
+    #[allow(useless_let_if_seq)]
+    fn record_handle(&mut self, resp: Option<&mut Response>, mut exec_metrics: ExecutorMetrics) {
+        let handle_start = self.handle_start.take().unwrap();
+        let now = Instant::now_coarse();
+        self.handle_time = Some(duration_to_sec(now - handle_start));
+        self.wait_start = Some(now);
+        self.total_handle_time += self.handle_time.unwrap();
+
+        self.exec_metrics.merge(&mut exec_metrics);
+
+        let mut record_handle_time = self.record_handle_time;
+        let mut record_scan_detail = self.record_scan_detail;
+        if self.handle_time.unwrap() > SLOW_QUERY_LOWER_BOUND {
+            record_handle_time = true;
+            record_scan_detail = true;
+        }
+        if let Some(resp) = resp {
+            if record_handle_time {
+                let mut handle = HandleTime::new();
+                handle.set_process_ms((self.handle_time.unwrap() * 1000f64) as i64);
+                handle.set_wait_ms((self.wait_time.unwrap() * 1000f64) as i64);
+                resp.mut_exec_details().set_handle_time(handle);
+            }
+            if record_scan_detail {
+                let detail = self.exec_metrics.cf_stats.scan_detail();
+                resp.mut_exec_details().set_scan_detail(detail);
+            }
+        }
+    }
+}
+
+impl Drop for RequestTracker {
+    fn drop(&mut self) {
+        COPR_PENDING_REQS
```

Comment:
Seems the wrong place to dec pending reqs. It should be done right before the request is handled.